✨  feat : 체험 프로그램 전체 조회 API 구현 및 관련 설정 추가

j-hann opened the pull request:
🟪 ProgramController 체험 프로그램 전체 조회 API 구현
  - `@ModelAttribute`로 `ProgramSearchRequestDto` 매핑하여 필터 및 정렬 조건 처리
  - 응답 `PaginationResponseDto` 페이지네이션 적용

🟪 ProgramService `searchProgram()` 메서드 구현
  - 조회된 `Program` List를 `ProgramCreateResponseDto`로 변환 후 `PaginationResponseDto`로 감싸서 리턴

🟪 ProgramRepositoryCustom & ProgramRepositoryCustomImpl 구현
  -` QueryDSL` 사용하여 동적으로 `BooleanExpression` 조건 생성
  - `지역` & `적성 카테고리 ID` 목록에 따라 동적 `where절` 구성
  - `정렬 기준(SortType)`을 `OrderSpecifier`로 매핑하여 `orderBy`에 적용
  - `getSortOrder()` 메서드 타입 안정성을 위해 `OrderSpecifier<? extends Comparable<?>>`로 제네릭 타입 명시, `Comparable` 구현 타입만 허용
    - `sortType` `null`인 경우, 최신순으로 정렬
  - `offset` & `limit` 방식으로 페이지네이션 처리
  - 정렬 타입이 없거나 잘못된 경우 예외 처리

🟪 ProgramSearchRequestDto 구현
  - 페이지 번호, 페이지 크기 디폴트값 설정
  - `@ModelAttribute` 바인딩 위해 `@Setter` 추가, 필드값 자동 주입 가능하도록 처리

🟪 PaginationResponseDto 공통 DTO 구현
  - `데이터 List`, `현재 페이지 번호`, `페이지 크기`, `전체 항목 수`, `전체 페이지 수` 필드 추가
  - `Page<T>`에서 `PaginationResponseDto<T>`로 변환하는 `of()` 정적 팩토리 메서드 추가

🟪 SortType Enum 추가
  - 타입 안정성 위해 정렬 옵션을 Enum으로 정의
  - `LATEST`, `PRICE_ASC`, `PRICE_DESC`, `REVIEW` 상수값 추가

🟪 QuerydslConfig 추가
  - `JPAQueryFactory`를 `Bean`으로 등록해 `Repository`에서 주입하여 사용할 수 있도록 설정

🟪 `build.gradle` QueryDSL 관련 의존성 추가
  - Q클래스를 `src/main/generated` 하위에 생성되도록 빌드 설정 추가
  - `JavaCompile`, `sourceSets`, `clean` 설정 추가, 빌드시 Q파일 자동 포함 및 정리

🟪 ErrorCode 추가

🟪 program.http 스크립트 추가
  - 체험 프로그램 전체 조회 API 테스트 http 요청 코드 추가

🟪 **TODO**
- _리뷰 별점 관련 조회는 추후 추가 예정_
- _제목 검색 로직도 추후 추가 예정_

j-hann committed:
✨  feat : 체험 프로그램 전체 조회 API 구현 및 관련 설정 추가

- ProgramController 체험 프로그램 전체 조회 API 구현
  - @ModelAttribute로 ProgramSearchRequestDto 매핑하여 필터 및 정렬 조건 처리
  - 응답 PaginationResponseDto 페이지네이션 적용

- ProgramService searchProgram() 메서드 구현
  - 조회 결과 ProgramCreateResponseDto로 변환 후 PaginationResponseDto로 감싸서 리턴

- ProgramRepositoryCustom & ProgramRepositoryCustomImpl 구현
  - QueryDSL 사용하여 동적으로 BooleanExpression 조건 생성
  - 지역 & 적성 카테고리 ID 목록에 따라 동적 where절 구성
  - 정렬 기준(최신순, 가격 오름차순, 가격 내림차순)을 OrderSpecifier로 매핑하여 orderBy에 적용
  - offset & limit 방식으로 페이지네이션 처리
  - 정렬 타입이 없거나 잘못된 경우 예외 처리

 - ProgramSearchRequestDto 구현
   - 페이지 번호, 페이지 크기 디폴트값 설정
   - @ModelAttribute 바인딩 위해 @Setter 추가, 필드값 자동 주입 가능하도록 처리

- PaginationResponseDto 공통 DTO 구현
  - 데이터 List, 현재 페이지 번호, 페이지 크기, 전체 항목 수, 전체 페이지 수 필드 추가
  -Page<T>에서 PaginationResponseDto<T>로 변환하는 of() 정적 팩토리 메서드 추가

- SortType Enum 추가
  - 타입 안정성 위해 정렬 옵션을 Enum으로 정의
  - LATEST, PRICE_ASC, PRICE_DESC, REVIEW 상수값 추가

- QuerydslConfig 추가
  - JPAQueryFactory를 Bean으로 등록해 Repository에서 바로 사용할 수 있도록 설정

- build.gradle QueryDSL 관련 의존성 추가
  - Q클래스를 src/main/generated 하위에 생성되도록 빌드 설정 추가
  - JavaCompile, sourceSets, clean 설정 추가, 빌드시 Q파일 자동 포함 및 정리

- ErrorCode 추가

- program.http 스크립트 추가

j-hann committed:
🔀  branch : develop merge

 branch : develop merge

j-hann committed:
🔥  remove : 불필요한 의존성 삭제

build.gradle 필요없는 의존성 삭제

j-hann committed:
🔧  add : build.gradle 의존성 추가

- Github Actions 빌드 실패로 인해 Q타입 생성 관련 의존성 추가

jiyoon0000 commented:
지연님,
QueryDSL에서 자동으로 생성되는 Q 클래스들은 `.gitignore`에 `build/generated/` 경로를 포함시켜서 제외해야 하는 것으로 알고 있습니다!

이 파일들은 빌드 시점에 항상 자동 생성되기 때문에 깃에 포함될 필요가 없다고 해요....🥲
(지금 방향대로 하면 엔티티 변경때마다 불필요한 변경사항이 자꾸 쌓이게 됩니다.🤯)

혹시 가능하다면 제외하고 다시 PR 올려주실 수 있을까요?🥹

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/program/controller/ProgramController.java:
```diff
@@ -74,4 +78,33 @@ public ResponseEntity<CommonResponseDto<ProgramCreateResponseDto>> createProgram
+	@Operation(
+		summary = "체험 프로그램 조회",
+		description = "유저가 정렬 조건을 설정하여 체험 프로그램을 조회 할 수 있습니다.",
+		security = @SecurityRequirement(name = "bearerAuth")
```
`SecurityConfig`에서 이미 모든 API가 JWT 인증을 거치도록 설정되어 있고 `SwaggerConfig`에서도 전역으로 `bearerAuth`가 설정되어 있어서 이 메타데이터는 중복이 되는 느낌이라....안쓰셔도 될 것 같아요!

jiyoon0000 replied:
위에 체험프로그램 생성 API에서도 지우셔도 될 것 같아요

j-hann replied:
넵 수정했습니다~

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/program/repository/ProgramRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,93 @@
+	/**
+	 * 정렬 조건 설정
+	 */
+	private OrderSpecifier<?> getSortOrder(SortType sortType, QProgram program) {
```
`program.createdAt`, `program.finalPrice` 둘 다 `Comparable` 타입이라 와일드 카드 대신 `<Comparable>`로 명확히 지정하면 좋을 것 같습니다.

j-hann replied:
필드마다 타입이 달라서 와일드 카드로 해놨는데 정렬 필드들이 Comparable 구현한 타입이니까 명시해두면 좋겠네요! 수정했습니다!👍

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/program/repository/ProgramRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,93 @@
+	private OrderSpecifier<?> getSortOrder(SortType sortType, QProgram program) {
+		OrderSpecifier<?> orderSpecifier;
+
+		if (sortType == null) {
```
현재 sortType이 null이면 바로 `INVALID_SORT_TYPE` 예외를 던지고 있는데, 이렇게 되면 프론트에서 sort 파라미터를 무조건 보내야해서 기본적으로 최신순 조회가 되도록 디폴트 값을 넣어주는게 어떨까요?

j-hann replied:
디버깅하다가 디폴트 설정 로직 삭제하고 다시 추가하는 걸 잊어버렸네여.. 감사합니다!!!👍

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/program/controller/ProgramController.java:
```diff
@@ -74,4 +78,33 @@ public ResponseEntity<CommonResponseDto<ProgramCreateResponseDto>> createProgram
+
+		PaginationResponseDto<ProgramCreateResponseDto> responsePage = programService.searchProgram(requestDto, memberId);
+
+		PaginationResponseDto<ProgramCreateResponseDto> pagination = new PaginationResponseDto<ProgramCreateResponseDto>(
```
현재 `ProgramService` 코드에서 `return PaginationResponseDto.of(programContent);` 이미 dto 형태로 반환하고 있어서, Controller에서 `new PaginationResponseDto<>` 로 감싸는 부분은 불필요할 것 같아요.

그냥 Service return 값을 바로 사용하면 더 깔끔할 것 같습니다.

j-hann replied:
넵 수정했습니다~

j-hann committed:
♻️  refactor : ProgramController 중복 처리 제거, ProgramRepositoryCustomImpl 정렬 기본값 추가, getSortOrder() 제네릭 타입 명시

- ProgramController 중복 처리 제거, ProgramRepositoryCustomImpl 정렬 기본값 추가, getSortOrder() 제네릭 타입 명시

j-hann committed:
🔧  add : ProgramRepository & ReservationRepository 기본 메서드 추가

- Service에서 매번 에외처리 안하도록 Repository에 공통 메서드 정의

jiyoon0000 commented:
ProgramRepository -> PROGRAM_NOT_FOUNT
ReservationRepository -> RESERVATION_NOT_FOUNT

두개가 ErrorCode enum에 정의되어있지 않아서 CI run 실패가 뜨는 것 같습니다!

j-hann commented:
> ProgramRepository -> PROGRAM_NOT_FOUNT ReservationRepository -> RESERVATION_NOT_FOUNT
> 
> 두개가 ErrorCode enum에 정의되어있지 않아서 CI run 실패가 뜨는 것 같습니다!

j-hann committed:
🔧  add : ErrorCode 추가

- Program, Reservation 관련 ErrorCode 추가
